### ISSUE TITLE
Implement einsum function/macro à la PyTorch and TF

### ISSUE THREAD

**Issue by outlace** [FluxML/Flux.jl#297]
[ Opening a new issue as discussed in closed issue #292 ]

_Issue_: PyTorch and TensorFlow both implement a function to compute tensor contractions using einstein summation notation (the einsum function). It would be great if Flux would also have this feature.  [This PyTorch commit](https://github.com/wsttiger/pytorch/commit/e8b860eee157bba2b8188632b66d82bc78017224) shows how einsum was implemented there. It seems it can be done just by reshaping the tensors into matrices and permuting indices, then reshaping back into higher order tensors.

_Background_: Tensor contractions generalize the various kinds of tensor products, e.g. matrix-matrix multiplications, matrix-vector multiplications, trace,  dot products, outer products, etc.  For example, if you want to multiply two matrices A and B using Einstein summation notation, you first take their tensor product A x B, which forms a 4th-order tensor (i.e. a tensor with 4 indices). Then you label each index of this tensor with arbitrary labels such as "a,b,b,c." Notice the 2nd and 3rd indices use the same label, so this implies we contract (or sum) over this pair of indices (the indices must be the same size for this to work) resulting in a new tensor C which only has 2 indices (a matrix). See [here](https://physics.stackexchange.com/questions/252473/converting-between-matrix-multiplication-and-tensor-contraction) for more math details of the computation.

_Motivation_: Beyond the convenience of using a single function to implement these various products, you can use tensor contractions to implement tensor networks, which are like deep neural networks except they utilize higher order tensor products and do not generally involve an explicit non-linear function application, and yet they can be used for prediction, classification, and data compression of high-dimensional data. 

_Possible Syntax_: The TensorOperations.jl package implements a nice macro for einstein notation that is different than PyTorch or TF. A contraction of two matrices is represented as `@tensor C[a,c] = A[a,b]*B[b,c]`. In PyTorch (and TF and NumPy) this would be `torch.einsum('ab,bc->ac', (A,B))`. Another possible syntax could be `contract(A,B,(1,2,2,3))` where the first two arguments are the tensors to be contracted and the third argument is a tuple of indices labeled as integers, although this is a bit less expressive.

**Comment by darsnack** [comment]
I think this can be closed. Anyone wanting this feature can use Tullio.jl. But if you want it as part of Flux, then open a new issue about reexporting Tullio.jl + friends.

**Comment by DhairyaLGandhi** [comment]
What's wrong with this one?

**Comment by darsnack** [comment]
See [here](https://github.com/FluxML/Flux.jl/pull/299#issuecomment-777941501).

**Comment by CarloLucibello** [comment]
I don't think there will ever need to have this in Flux. Pytorch has to offer this functionality for his tensors, but in Flux we work with standard arrays, which are egregiously covered by Tullio.jl

**Comment by mcabbott** [comment]
I don't think Flux is the right place to implement such things. I think it's a limitation of Python that TF & Pytorch & Jax all have to re-implement their own near-clones of Numpy's `einsum`, each for slightly different array types (& autodiff engines).

While I'm pleased that people like Tullio.jl so much, the closest analogue is probably [OMEinsum.jl](https://github.com/under-Peter/OMEinsum.jl). This (mostly) acts by "reshaping the tensors into matrices and permuting indices" as suggested above, and can accept Numpy-like notation `ein"ab,bc->ac"(A,B)`. It should handle autodiff and GPU arrays. 

If someone wanted to revive the simpler approach of #299, the right place to do it would be a new package. Such a thing might be useful independent of Flux, there's no need for them to be coupled.
